File: Java/src/main/java/com/phasmidsoftware/dsaipg/adt/trie/TrieBenchmark.java
package com.phasmidsoftware.dsaipg.adt.trie;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;

/**
 * The TrieBenchmark class is used to benchmark the performance of various operations
 * like inserting words, searching for prefix matches, and deleting words using a Trie (prefix tree) data structure.
 * It also compares the performance of Trie-based prefix searches with a linear search approach.
 * <p>
 * The benchmarking involves measuring the execution time for:
 * - Inserting words into a Trie.
 * - Searching for auto-complete suggestions based on a given prefix using the Trie.
 * - Performing a linear prefix search on a list of words.
 * - Deleting words from the Trie.
 *
 * @author Shivani Sugurushetty (https://github.com/shivanisugurushetty)
 * (<a href="https://github.com/shivamsugurushetty/DSAIPG">DSAIPG (fork)</a>)
 */
public class TrieBenchmark {

    /**
     * The main method serves as the entry point for the application and benchmarks the performance
     * of various operations related to the Trie data structure. These operations include:
     * - Inserting words into the Trie.
     * - Searching for auto-complete results based on a prefix using the Trie.
     * - Comparing the search performance of the Trie with a linear search.
     * - Deleting words from the Trie.
     * <p>
     * TODO use Stopwatch instead of System.nanoTime()
     *
     * @param args command-line arguments (not used in this implementation).
     *             The method expects no arguments for its execution.
     * @throws IOException if there is an error reading the dictionary file from the file system.
     */
    public static void main(String[] args) throws IOException {
        List<String> words = Files.readAllLines(Paths.get("src/main/resources/dictionary.txt"));

        Trie trie = new Trie();
        long start = System.nanoTime();
        for (String word : words) trie.insert(word);
        long trieInsertTime = System.nanoTime() - start;

        start = System.nanoTime();
        List<String> result = trie.autocomplete("ap");
        long trieSearchTime = System.nanoTime() - start;

        start = System.nanoTime();
        List<String> linearResults = new ArrayList<>();
        for (String word : words) if (word.startsWith("ap")) linearResults.add(word);
        long linearSearchTime = System.nanoTime() - start;

        start = System.nanoTime();
        for (String word : words) trie.delete(word);
        long trieDeleteTime = System.nanoTime() - start;

        System.out.printf("Trie insert time: %d ns%n", trieInsertTime);
        System.out.printf("Trie search time: %d ns%n", trieSearchTime);
        System.out.printf("Linear search time: %d ns%n", linearSearchTime);
        System.out.printf("Trie delete time: %d ns%n", trieDeleteTime);
    }
}

File: Python/src/adt/trie/trie_benchmark.py
import os
import time

from .trie import Trie


def main():
    """
    Benchmark the Trie implementation against linear search and measure performance of operations.
    """
    # Path to the dictionary file in the Java resources directory
    # Determine path relative to this script: .../Python/src/adt/trie/trie_benchmark.py
    current_dir = os.path.dirname(os.path.abspath(__file__))
    project_root = os.path.abspath(os.path.join(current_dir, "../../../../"))
    resource_path = os.path.join(project_root, "Java/src/main/resources/3000-common-words.txt")
    
    if not os.path.exists(resource_path):
        print(f"Error: Dictionary file not found at {resource_path}")
        return

    try:
        with open(resource_path) as f:
            words = [line.strip() for line in f if line.strip()]
    except Exception as e:
        print(f"Error reading dictionary file: {e}")
        return

    print(f"Loaded {len(words)} words from dictionary.")

    trie = Trie()

    # Benchmark Insert
    start = time.perf_counter_ns()
    for word in words:
        trie.insert(word)
    trie_insert_time = time.perf_counter_ns() - start

    # Benchmark Autocomplete
    prefix = "ap"
    start = time.perf_counter_ns()
    # Ignoring result for benchmark, just timing
    trie.autocomplete(prefix)
    trie_search_time = time.perf_counter_ns() - start

    # Benchmark Linear Search
    start = time.perf_counter_ns()
    # Linear search equivalent
    _ = [word for word in words if word.startswith(prefix)]
    linear_search_time = time.perf_counter_ns() - start

    # Benchmark Delete
    start = time.perf_counter_ns()
    for word in words:
        trie.delete(word)
    trie_delete_time = time.perf_counter_ns() - start

    print(f"Trie insert time: {trie_insert_time} ns")
    print(f"Trie search time: {trie_search_time} ns")
    print(f"Linear search time: {linear_search_time} ns")
    print(f"Trie delete time: {trie_delete_time} ns")

if __name__ == "__main__":
    main()
